### PR TITLE
Fix where clauses not taking the width of the line into account

### LINKED ISSUE #630: Long function declarations do not get line breaks correctly
Before:
```
pub trait Test {
    fn very_long_method_name<F>(self, f: F) -> MyVeryLongReturnType
        where F: FnMut(Self::Item) -> bool;
}
pub trait Test {
    fn very_long_method_name<F>(self, f: F) -> MyVeryLongReturnType
        where F: FnMut(Self::Item) -> bool
            , F: Fn();
}
```
After:
```
pub trait Test {
    fn very_long_method_name<F>(self, f: F) -> MyVeryLongReturnType where F: FnMut(Self::Item) -> bool;
}
pub trait Test {
    fn very_long_method_name<F>(self, f: F) -> MyVeryLongReturnType
        where F: FnMut(Self::Item) -> bool,
              F: Fn();
}
```

Might be part of #78 (?) but I think this may be a separate issue as adding an extra element to the `where` list will make it format correctly.

### PR DESCRIPTION
**Pull request by Marwes**
Fixes #630

### CODE REVIEW

**Comment by Marwes** [comment]
#650 is relevant I guess, I followed the current behavior which seems to assume that braces are always on a new line if a where clause exists.

File: src/items.rs [hunk 1]
```diff
@@ -1462,9 +1471,15 @@ fn rewrite_where_clause(context: &RewriteContext,
     };
     let preds_str = try_opt!(write_list(&item_vec, &fmt));
 
-    // 9 = " where ".len() + " {".len()
+    // When '{' is the terminator just assume that it is put on the next line for single line where
+    // clauses
```
Comment (nrc) [inline src/items.rs]: Is this ever not the case? I'd rather we are sub-optimal by one char than even occasionally over-shoot by a char.

**Comment by nrc** [comment]
Looks good, thanks!

**Comment by vrieskl** [comment]
Does this also address https://github.com/rust-lang-nursery/rustfmt/issues/654?

**Comment by Marwes** [comment]
Tried to use the item_brace_style instead since that may change where the brace is put but there is always fn_brace_style so that would need to be taken into account for functions.

The other alternative I am seeing is to always assume 2 chars needed when there is a brace at the end but that breaks tests which would barely fit items on a single line with the brace on the next line.

**Comment by Marwes** [comment]
@vrieskl  #654 seems to be the same problem as #630 so they should both be fixed if I can get everything right.

**Comment by marcusklaas** [comment]
Sorry for leaving you without feedback for so long. Shouldn't we pass the brace style as parameter so we can use `fn_brace_style` and `item_brace_style` when applicable? Other than that, this is good to go!

**Comment by Marwes** [comment]
I don't think this is the best way of dealing with it as there is a need to pass `brace_style` even when the terminator is ';' or '=' (working on a fix for type aliases as well) but it works and I don't have a better idea for how to handle it currently so merging this should be fine now.

**Comment by marcusklaas** [comment]
Thanks for this, @Marwes!

**Pull request merged**